### ISSUE TITLE
Bug: [no-unsafe-type-assertion] Fails to report unsafe assertions involving constrained type parameters

### ISSUE THREAD

**Issue by controversial** [typescript-eslint/typescript-eslint#10453]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link
[Playground link](https://typescript-eslint.io/play/#ts=5.5.4&fileType=.tsx&code=C4TwDgpgBA8gRgKwgY2FAvFA3gKClAMwHsiAuKAZ2ACcBLAOwHMBuHAXx04HoupGSAJlDgQAFgEMAbrSLVC1IgFsooSBWR0wwUjggAPMLLQEArvVQz6hM8gA8AYRNUl8JKgAqUfcAj0BFWEQUYAA%2BAAoiRHJHZ0VXYPcASminYBcgj2w8KB4oMIEiSiUIYFEGRigAd1pSlVFoBjATYETsySJaIUiEVnxcjQhxHyhxKHoISqhu4MJZOugiZqbgbOQieioxifjUch20TCxZsigAcnqQU6g2XpzeAHEYGAARclLaAIJxWgAbAJFkOInNB9p41iYfkIRFBFLJoBRIMhaARaMg6uIrJUJGhKtABkMIAJVutNuNKjE0nEMtooBT0m5gJ5MGT9rdqCUTNQrGS6VSGawONxeHBxFCxFIZHICAplPQiABaMwUcQECDy1Rq8QUCgQajASw6fSGPXWcz69am5AAJgcqXpCS8eh8fgC%2B3C3RSsVByVpdr5DtwfV4%2BUKFGK7yYVRqonmUEazVa%2BHanSmiFu-XZBJGW0m01QszkpQWS2axI2aBZ1L21Iw2GO5HOEEu13TvAAQgBBV51D5bQkBNJQcH0NBaqCAHAIlSroBqRtrdeb6IBcAjjBBG9BAUEk4h%2BJgWa6LKnA0DA4mo4kUJV1VUWkJh4gA1g00ErFn4y6SJrz9rXKwy576XrUu42TssAnLcl%2BfqsuwOBAA&eslintrc=N4KABGBECmAeAu0B2ATAzpAXGA2uCUADgDYCuA5gJZKYAC8AnodGgMYBOlh8AtC8dXgB6dtFYB7ALaTkKaCh6NmPVgAsxAa3mQANPgiQSFanSUsOXXv0FC0jAXcqtFTaCvWstKSPgC6eg3ZSYhYsMFACKHpXNk5uPjQBJGEJJDRKO2ReMx45ADNqSnhKcTSwnBh2dnF2XSgzSH99KLNYywSk4SRxHlI0gEM8txz%2BtDRodmLSsMrq2vwAXxAFoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg2RGwAqkWgALdNADW6ACYBJIjPQAPWQEFo0dCTKUO6XgF8QhoA&tokens=false)

### Repro Code

```TypeScript
type Object = {
  foo: string;
}

export function myfunc<CustomObjectT extends Object>(input: CustomObjectT): CustomObjectT {
  const newObject: Object = { foo: 'hey' };
  // This is UNSAFE because CustomObjectT can/will be instantiated with a narrower subtype of Object
  // no-unsafe-type-assertion should (but doesn’t) catch this
  const newCustomObject = newObject as CustomObjectT;
  return newCustomObject;
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unsafe-type-assertion": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

The line with the type assertion involving the generic parameter (`newObject as CustomObjectT`) should report `@typescript-eslint/no-unsafe-type-assertion` by the same logic wherein Typescript considers such a conversion unsafe:
> ```
> 2322: Type 'Object' is not assignable to type 'CustomObjectT'.
>   'Object' is assignable to the constraint of type 'CustomObjectT', but 'CustomObjectT' could be instantiated with a different subtype of constraint 'Object'.
> ```

The first example in the playground shows that this case is not an allowable *assignment* according to TS, which, to my understanding, should be the heuristic used by `no-unsafe-type-assertion`

### Actual Result

No error is reported from `@typescript-eslint/no-unsafe-type-assertion`

### Additional Info

cc @ronami

thanks for your hard work on this rule!

**Comment by ronami** [comment]
Interesting!

I think it is a valid case the rule should warn on, as this example literally narrows down a type from its type constraint to a potentially more specific one.

I initially thought this has to do with [type-widening](https://github.com/typescript-eslint/typescript-eslint/pull/10051#discussion_r1800019322), but it looks like it's related to how type constraints are handled. The rule compares the types (or the type constraints if the compared types have any). Since both have the same type constraint, [the validation stops early on](https://github.com/typescript-eslint/typescript-eslint/blob/c92c29d142640b107795cfcdd28e94e4a811872d/packages/eslint-plugin/src/rules/no-unsafe-type-assertion.ts#L61-L63). I think a potential solution would be to check the specific types, rather than the type constraints if both type constraints are the same.

**Comment by controversial** [comment]
Ok, interesting!

So to clarify my understanding of what you said: this behavior is because the rule obtains the `expressionType` and `assertedType` using `getConstrainedTypeAtLocation`, which “[will resolve to the type's generic constraint, if it has one](https://github.com/typescript-eslint/typescript-eslint/blob/c92c29d142640b107795cfcdd28e94e4a811872d/packages/type-utils/src/getConstrainedTypeAtLocation.ts#L8)” ?

Which means the rule will need to have logic for the special case of generic constraints—because it’s **not** generally true that *A is assignable to B if A’s generic constraint is assignable to B’s* (since B “could be instantiated with a different subtype of constraint”)

**Comment by kirkwaiblinger** [comment]
Huh! We're so used to checking generics' constraints, but, yeah, that might actually be a mistake for this rule.

**Comment by kirkwaiblinger** [comment]
Might be nice if we can special-case the error message here since the TS error is one that always surprises and confuses me at first when I run into it.

**Comment by controversial** [comment]
What do you think a better error message would be?

**Comment by kirkwaiblinger** [comment]
> What do you think a better error message would be?

As a first pass, I think we could just model it after the TS error message?

```text
Unsafe type assertion: 'Object' is assignable to the constraint of type 'CustomObjectT', but 'CustomObjectT' could be instantiated with a different subtype of constraint 'Object'.
```

Doesn't necessarily have to be that, verbatim 🤷... if we get working code that detects this case as the reason for erroring, we could probably bikeshed about wording on the PR. 

The only import technical thing here is ensuring that we _don't_ give this error message on a case like 
```ts
export function myfunc<CustomObjectT extends string>(input: number): CustomObjectT {
  const newCustomObject = input as CustomObjectT;
  return newCustomObject;
}
```
That snippet _is_ unsafe, but because `number` is not assignable to `string` at all, not because of the generic subtypes thing.  (Note that it's also a TS error in this case, so it's not super important, but I'm guessing we could come up with a corollary that's not a TS error)

**Comment by controversial** [comment]
I started taking a stab at this in #10461—but I really don’t know what I’m doing, so I’d welcome feedback!

**Comment by controversial** [comment]
> The only import technical thing here is ensuring that we don't give this error message on a case like
> ```ts
> export function myfunc<CustomObjectT extends string>(input: number): CustomObjectT {
>   const newCustomObject = input as CustomObjectT;
>   return newCustomObject;
> }
> ```

I added this example as a test case to ensure we don’t give the “assignable to the constraint” message here!
